File: Problem_1.py
"""
Time Complexity : O(hw c n) 
Space Complexity : O(hw)- for the matrix
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No

Your code here along with comments explaining your approach:
To me, this problem seems a combination of 2 problems. The N queens problem , where we determine the correct placement and the
01 matrix problem where we find the minimum  distance for the nearest 0. Here also, we make a grid of given height and width
and pass it to a backtrack function to find the correct placement of buildings. To do that, we try all the combinations, and for 
every combination, we perform bfs for finding out the minimum distance from the farthest parking lot. 
"""
from collections import deque


class Solution:

    def __init__(self):
        self.minDistance = float('inf')
        self.dirs = [(-1, 0), (1, 0), (0, 1), (0, -1)]

    def optimalPlacement(self, h, w, n):
        grid = [[-1 for i in range(w)] for j in range(h)]
        self.backtrack(grid, 0, 0, h, w, n)
        return self.minDistance

    def backtrack(self, grid, r, c, h, w, n):
        if n == 0:
            self.bfs(grid, h, w)
            return

        if c == w:
            r += 1
            c = 0

        for i in range(r, h):
            for j in range(c, w):
                # action
                grid[i][j] = 0
                # recurse
                self.backtrack(grid, r, c+1, h, w, n-1)
                # backtrack
                grid[i][j] = -1
            c = 0

    def bfs(self, grid, h, w):
        q = deque([])
        visited = [[False for i in range(w)] for j in range(h)]
        for i in range(h):
            for j in range(w):
                if grid[i][j] == 0:
                    q.append([i, j])
                    visited[i][j] = True
        distance = 0
        while q:
            size = len(q)
            for i in range(size):
                curr = q.popleft()
                for d in self.dirs:
                    r = curr[0]+d[0]
                    c = curr[1]+d[1]
                    if r >= 0 and c >= 0 and r < h and c < w and not visited[r][c]:
                        q.append([r, c])
                        visited[r][c] = True
            distance += 1
        self.minDistance = min(self.minDistance, distance-1)


def main():
    s = Solution()
    print(s.optimalPlacement(4, 4, 2))


if __name__ == '__main__':
    main()

File: Problem1.java
// Time Complexity :O(n!)
// Space Complexity :O(n*n)
// Did this code successfully run on Leetcode :es
// Any problem you faced while coding this :No


// Your code here along with comments explaining your approach
/*package whatever //do not write package name here */

import java.io.*;
import java.util.*;

class GFG {
	public static void main (String[] args) {
		BuildingPlacement obj = new BuildingPlacement();
		System.out.println(obj.findDistance(3,3,2));
	}
}
    class BuildingPlacement{
    int minDistance= Integer.MAX_VALUE;
    
    public int findDistance(int h, int w, int b){
        int[][] grid= new int[h][w];
        
        backtrack(grid,0,0,b,h,w);
        return minDistance;
    }
    int[][] dirs= new int[][]{{-1,0},{1,0},{0,1},{0,-1}};
    public void dfs(int[][]grid, int h, int w){
        Queue<int[]> q= new LinkedList<>();
        boolean[][] visited= new boolean[h][w];
        for(int i=0;i<h;i++){
            for(int j=0;j<w;j++){
                if(grid[i][j]==1){
                    q.add(new int[]{i,j});//adding  all the placed buildings
                    visited[i][j]=true;
                }
            }
        }
        int level=0;//finding the paths which are farthest from building placed
        while(!q.isEmpty()){
            int size= q.size();
            for(int x=0;x<size;x++){
                int[] front = q.poll();
                int i=front[0]; int j= front[1];
                for(int[] dir: dirs){
                    int r=i+dir[0];
                    int c=j+dir[1];
                    if(r<h && c<w && r>=0 && c>=0 && !visited[r][c]){
                        q.add(new int[]{r,c});
                        visited[r][c]=true;;
                    }
                }
            }
            level++;
        }
        minDistance= Math.min(minDistance,level-1);
        
    }
    public void backtrack(int[][] grid, int r, int c, int b, int h, int w){//all combinations of building placed
        if(b<=0){
            dfs(grid,h,w);
            return;
        }
        if(c>=w){//we are moving in each from left to right and then increase row
            r++;
            c=0;
        }
        for(int i=r;i<h;i++){
            for(int j=c;j<w;j++){
                grid[i][j]=1;//placing the building at i,c
                backtrack(grid,i,j,b-1,h,w);
                grid[i][j]=0;//removing the building at i,c
            }
        }
        
        
    }
    
}